wasm/wat/types.go

```diff
@@ -0,0 +1,83 @@
+package wat
+
+import "fmt"
+
+// module corresponds to the text format of a WebAssembly module, and is an intermediate representation prior to
+// wasm.Module. This is primarily needed to resolve symbolic indexes like "$main" to raw numeric ones.
+//
+// Note: nothing is required per specification. Ex `(module)` is valid!
+//
+// See https://www.w3.org/TR/wasm-core-1/#functions%E2%91%A7
+type module struct {
+	// name is optional and starts with '$'. For example, "$test".
+	// See https://www.w3.org/TR/wasm-core-1/#modules%E2%91%A0%E2%91%A2
+	//
+	// Note: The name may also be stored in the wasm.Module CustomSection under the key "name" subsection 0.
+	// See https://www.w3.org/TR/wasm-core-1/#binary-namesec
+	name string
+
+	// imports are added in insertion order.
+	imports []*_import
+
+	// startFunction is the function to call during wasm.Store Instantiate. The value is a importFunc.name, such as
+	// "$main", or its equivalent raw numeric index, such as "2".
+	//
+	// Note: When in raw numeric form, this is relative to imports. See wasm.Module StartSection for more.
+	//
+	// See https://www.w3.org/TR/wasm-core-1/#start-function%E2%91%A4
+	startFunction string
+}
+
+// importFunc corresponds to the text format of a WebAssembly function import description.
+//
+// Note: nothing is required per specification. Ex `(func)` is valid!
+//
+// See https://www.w3.org/TR/wasm-core-1/#imports%E2%91%A0
+type importFunc struct {
+	// name is optional and starts with '$'. For example, "$main".
+	//
+	// This name is only used for debugging. At runtime, functions are called based on raw numeric index. The function
+	// index space begins with imported functions, followed by any defined in this module.
+	// See https://www.w3.org/TR/wasm-core-1/#functions%E2%91%A7
+	//
+	// Note: The name may also be stored in the wasm.Module CustomSection under the key "name" subsection 1.
+	// See https://www.w3.org/TR/wasm-core-1/#binary-namesec
+	name string
+
+	// TODO: typeuse https://www.w3.org/TR/wasm-core-1/#text-typeuse
+}
+
+// _import corresponds to the text format of a WebAssembly import.
+//
+// See https://www.w3.org/TR/wasm-core-1/#imports%E2%91%A0
+type _import struct { // note: this is named _import because import is reserved in golang
```

Comment:
How about?

```suggestion
type importField struct { // note: this is named _import because import is reserved in golang
```

Reply:
only thing is we use the word import field to describe a field in the import ex func. ex importFieldHandler handles func

Reply:
once I had `moduleImport` but I reverted it. exporting was the other way to avoid `_`, but then also raises another question. it is a crappy thing that we can name a string variable "string", but we can't name a type "import" :p

Reply:
ok I found another way out, which is to group import types together. For example, `importFunc` doesn't clash with the keyword `import``